pass_args_to_sdf.md

```diff
@@ -0,0 +1,201 @@
+
+![ign-gazebo's Server current process of loading SDF](./images/ign_gazebo_curr_SDF_process.png)
+
+The figure above shows the current process, where the gazebo `Server` takes in an SDF file and libsdformat loads the SDF object. In libsdformat's `Load(filename)` function, `readFile` loads the file into `sdfParsed` which is an `SDFPtr` that contains the root `sdf::Element` as well as other information about the SDF file. Then the `Load(sdfParsed)` creates the DOM object.
+
+![ign-gazebo's Server proposed process of loading SDF](./images/ign_gazebo_new_SDF_process.png)
+
+The figure above shows the possible new implementation for the gazebo `Server` (shown in green). Where the server calls `readFile` to load `sdfParsed`. Then the contents of `sdfParsed` will be modified with the new updated parameters set by the user. Finally, `sdfRoot.Load(sdfParsed)` will be called to set up the DOM object.
+
+The reason for this solution is when a file is loaded to a `SDFPtr` then to a DOM object, it is possible that the DOM object and SDF element are not insync. Meaning, modifying the DOM object may not propagate these changes. Therefore the solution was to modify the `SDFPtr` before it is created into a DOM object. Also, the advantage of using the `SDFPtr` (i.e., `sdfParsed`) is that we can utilize libsdformat's API (e.g., [Element](http://osrf-distributions.s3.amazonaws.com/sdformat/api/9.0.0/classsdf_1_1v9_1_1Element.html) class) to update elements as opposed to using some other XML parser.
+
+### Parameter Specification
+
+Parameters will be specified by the user in either launch (`.ign`) or world (`.sdf`) files through the use of model [composition (or nested models)](http://sdformat.org/tutorials?tut=composition) which will contain `<include>` tags and [custom elements](http://sdformat.org/tutorials?tut=custom_elements_attributes_proposal&cat=pose_semantics_docs). Model composition provides a way to create a model by nesting other models within itself. Discussed in this section is the proposed design for specifying parameters and the actions available to the user. The intent is to provide a means of manipulating almost all parameters in SDF models through launch/world files without requiring modifications to already existing/created model files. There is a limitation with this proposed approach with specifying plugins since plugins may not have unique names. This will be elaborated later on as well as a potential method of resolving this will be discussed, which may require a user to slightly modify the original model.
```

Comment:
> There is a limitation with this proposed approach with specifying plugins since plugins may not have unique names.

I think this is actually violation of the SDFormat 1.7 spec by ign-gazebo, since the [SDFormat 1.7 spec requires sibling elements of any type to have unique names](http://sdformat.org/tutorials?tut=pose_frame_semantics_proposal&cat=pose_semantics_docs&#3-2-unique-names-for-all-sibling-elements). I believe `ign-gazebo` is currently specifying plugin class types using the `name` attribute, but it would be better to specify them with a custom namespaced attribute like:

~~~
<plugin name='unique_name' filename='...' ignition:class='ignition::gazebo::systems::SystemName' />
~~~

Reply:
https://github.com/ignitionrobotics/ign-gazebo/issues/399